src/Pirouette/SMT/Base.hs

```diff
@@ -0,0 +1,58 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+module Pirouette.SMT.Base where
+
+import Control.Monad.IO.Class
+import qualified Pirouette.SMT.SimpleSMT as SimpleSMT
+import Pirouette.Term.Syntax
+import Data.Void
+
+-- | Captures the languages that can be translated to SMTLIB; namelly,
+-- we need to be able to translate each individual base syntactical category.
+--
+-- This class is defined with @-XAllowAmbiguousTypes@ and therefore
+-- should be used with @-XTypeApplications@ whenever necessary.
+class (LanguageDef lang) => LanguageSMT lang where
+  translateBuiltinType :: BuiltinTypes lang -> SimpleSMT.SExpr
+  translateBuiltinTerm :: BuiltinTerms lang -> SimpleSMT.SExpr
+  translateConstant :: Constants lang -> SimpleSMT.SExpr
```

Comment:
According to the [docs on `-XAllowAmbiguousTypes`](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/ambiguous_types.html) the check is just delayed from the declaration of the class/constraint to the call-site.

Reply:
My first concern is that's a module-level pragma, so even when it's needed for one definition, it also affects all others in the module. Hence, ambiguities can slip in those other definitions, and they will only be caught at the usage point (or not caught at all, if that's a library — but that's not exactly our case here).
Secondly, I remember having certain issues with ambiguous types and rank-n polymorphism, and I could try reproducing that if you're curious.

Consider this to be just my gut instinct to avoid this pragma. No strong opinion here on my side.